### DISCUSSION TITLE
Add the ability to assign Roles to many Teams

### DISCUSSION THREAD

**Discussion by Xety** [spatie/laravel-permission#2508] [Ideas]
Hello,

On the doc, when you want to assign roles/permissions to a team it say : `The role/permission assignment and removal for teams are the same as without teams, but they take the global team_id which is set on login.`

Ok that is cool, but what if you want assign a role to many teams or even many roles to many teams ?

I suggest something like this :
```php
$user = User::find(1);
// assignRole(array|Collection|int|Role|string $roles, array|Collection|int $teams)
$user->assignRole(['Role1', 'Role2', 'Role3'], [1, 2, 3]) // Assign many roles to many teams

// Other possible formats :
$user->assignRole('Role1', [1, 2, 3]) // Assign one role to many teams
$user->assignRole(['Role1', 'Role2', 'Role3'], 1) // Assign many roles to one team
```
The first `assignRole` would generate the following entries :

| role_id| model_type | model_id | team_id |
| ---- | ---- | ---- | ---- |
| 1 | App\Models\User | 1 | 1 |
| 1 | App\Models\User | 1 | 2 |
| 1 | App\Models\User | 1 | 3 |
| 2 | App\Models\User | 1 | 1 |
| 2 | App\Models\User | 1 | 2 |
| 2 | App\Models\User | 1 | 3 |
| 3 | App\Models\User | 1 | 1 |
| 3 | App\Models\User | 1 | 2 |
| 3 | App\Models\User | 1 | 3 |

**Comment by parallels999** [comment]
It is not possible
https://github.com/spatie/laravel-permission/blob/2cc536247cad8a2bebbad114a389d9af02612f97/src/Traits/HasRoles.php#L146
Make your own custom trait, or a custom method

**Reply by Xety** [reply]
I know it's not possible atm, that's why it's a feature request

**Reply by parallels999** [reply]
It is not possible because it is a breaking change, not because it has not been implemented.

> feature request

Make a PR to master

**Comment by drbyte** [comment]
There are probably additional multiple-team kinds of functions that should be done this way, beyond just assignRole().
So, if you're considering making a PR, I'd suggest adding some new team-specific functions instead of altering existing method signatures.

For example, if you were to create a function within your own app to do this proposed new feature, you'd probably do something like this:

You could put this into your User model (whatever model you'd need to use it with)
```php
/**
 * @param  string|int|array|Role|Collection|\BackedEnum  $roles
 * @param int|array $teams
 */
function assignRolesToTeams($roles, $teams)
{
    if (! app(PermissionRegistrar::class)->teams) {
        return;
    }

    $teams = Arr::wrap($teams);

    $team_id_to_be_restored = getPermissionsTeamId();

    foreach ($teams as $team) {
        setPermissionsTeamId($team);
        $this->assignRole($roles);
    }

    setPermissionsTeamId($team_id_to_be_restored);
}
```

When working on a proposal, consider what additional team-related functions would be needed too.

**Reply by Xety** [reply]
Hello,

Thanks for your reply. Yes you're right, the best approch would be to have team-specific functions.
It's the first time I'm using Teams with spatie/permissions, so I will explore that in my app, and come back after.
But I think it's a good idea to have some native team functions in the package, since the package support the team function, you will need to use some team-specific functions at time in your application.
If you see the `laratrust` package, they support natively teams functions :
```php
/// Roles
$user->addRole($admin, $team);
$user->addRoles([$admin, $owner], $team);
$user->removeRole($admin, $team);
$user->removeRoles([$admin, $owner], $team);
$user->syncRoles([$admin, $owner], $team);

// Permissions
$user->givePermission($editUser, $team);
$user->givePermissions([$editUser, $manageUsers], $team);
$user->removePermission($editUser, $team);
$user->removePermissions([$editUser, $manageUsers], $team);
$user->syncPermissions([$editUser, $manageUsers], $team);

// Check
$user->hasRole('admin', 'my-awesome-team');
$user->hasRole(['admin', 'user'], 'my-awesome-team', true);

$user->rolesTeams; // See all user teams with roles PRETTY NICE FUNCTON
$user->rolesTeams()->where('name','my-awesome-team')->first(); // Check if user has any role relating to a team
```
Actually to do all that you have to do :
```php
$team_id_to_be_restored = getPermissionsTeamId();
// ...
setPermissionsTeamId($team);
// ...
setPermissionsTeamId($team_id_to_be_restored);
```

Having a seccond param for `$teams` would be the best approch imo.
Anyways, I will continue to explore the Teams in my app and come back later with maybe some teams-specific functions.

**Reply by drbyte** [reply]
> Having a second param for $teams would be the best approach imo.

... and that would require that (in one way or another) every one of those functions does that same backup/set/do/restore cycle inside each of those functions.
